### PR TITLE
Improve icons query

### PR DESCRIPTION
**Pull request by yabirgb**
This commit does:
1. Exit earlier with fewer checks if a custom icon is used
2. Spawn a greenlet to query the icon from coingecko
3. Notify with a 202 that the request has been processed

### CODE REVIEW

File: rotkehlchen/icons.py [hunk 1]
```diff
@@ -167,10 +176,8 @@ def get_icon(
         If not, all icons of the asset are queried from coingecko and cached
         locally before the requested data are returned.
         """
-        # First search custom icons
-        custom_icon_path = self.custom_iconfile_path(asset)
-        if custom_icon_path is not None:
-            return custom_icon_path
```
Comment (yabirgb) [inline rotkehlchen/icons.py:173]: I removed this since is duplicated logic from `asset_icon_path`

Comment (yabirgb) [inline rotkehlchen/icons.py:173]: I removed this since is duplicated logic from `asset_icon_path`

File: rotkehlchen/api/rest.py [hunk 2]
```diff
@@ -2921,8 +2927,17 @@ def get_asset_icon(
             )
             if response is not None:
                 return response
+            else:
+                return maybe_create_image_response(image_path=icon_path)
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:2931]: What are you doing here? Did you not already just check for the same icon path above?

File: rotkehlchen/icons.py [hunk 3]
```diff
@@ -127,7 +137,7 @@ def query_coingecko_for_icon(self, asset: Asset, coingecko_id: str) -> bool:
         # we only keep delisted asset coingecko mappings since historical prices
         # can still be queried.
         if asset.identifier in DELISTED_ASSETS:
-            self.failed_asset_ids.add(asset.identifier)
+            self.failed_asset_ids.add(asset.identifier, asset.identifier)
```
Comment (LefterisJP) [inline rotkehlchen/icons.py:140]: This is useless. Change it also everywhere else you use it. you are making a cache with double the needed size. You are caching the asset identifier both as the key and the value in the cache. For the value you don't need anything. Just a  boolean would be enough.

In fact this cache you are using is a key-value mapping cache. You don't need this. You need a set LRU cache.

File: rotkehlchen/icons.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/icons.py:173]: Do we have a test for cusom icon paths? Make sure we do if we don't already

File: rotkehlchen/icons.py [hunk 4]
```diff
@@ -167,10 +177,8 @@ def get_icon(
         If not, all icons of the asset are queried from coingecko and cached
         locally before the requested data are returned.
         """
-        # First search custom icons
-        custom_icon_path = self.custom_iconfile_path(asset)
-        if custom_icon_path is not None:
-            return custom_icon_path
+        if self.failed_asset_ids.get(asset.identifier) is not None:
```
Comment (LefterisJP) [inline rotkehlchen/icons.py:180]: you added a return value argument. Please edit the docstring to explain what this function returns.

File: rotkehlchen/api/rest.py [hunk 5]
```diff
@@ -2921,8 +2927,17 @@ def get_asset_icon(
...
 
-        image_path = self.rotkehlchen.icon_manager.get_icon(asset)
+        image_path, processed = self.rotkehlchen.icon_manager.get_icon(asset)
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:2933]: processed is not really the correct name to use here.

```suggestion
        image_path, launched_remote_query = self.rotkehlchen.icon_manager.get_icon(asset)
```

File: rotkehlchen/api/rest.py (hunk 2 above)
Comment (yabirgb) [reply rotkehlchen/api/rest.py:2931]: here is creating a the response but is true that adds a check that was already done. I created a new function to avoid that. Another system call saved

File: rotkehlchen/icons.py (hunk 1 above)
Comment (yabirgb) [reply rotkehlchen/icons.py:173]: yes, is in `test_upload_custom_icon`

File: rotkehlchen/utils/data_structures.py [hunk 6]
```diff
@@ -33,3 +33,32 @@ def remove(self, key: str) -> None:
     def clear(self) -> None:
         """Delete all entries in the cache"""
         self.cache.clear()
+
+
+class LRUSetCache(Generic[RT]):
+    """
+    LRU cache that works like a set.
+    Internally it uses an OrderedDict to keep the order that maps to None.
```
Comment (LefterisJP) [inline rotkehlchen/utils/data_structures.py:41]: I don't get it. Why did you use ordered dict internally? I don't understand what `"to keep the order that maps to None."` means

You probably mean: In order to be able to keep the order of insertion so that the LRU mechanism can function. Right?

Can you rephrase?

**Review by LefterisJP — APPROVED** [review]
LGTM. I left a nitpick comment on a docstring, but other than that, all green so I am merging.
